---
title: "diff_expression"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
dyn.load("/home/rstudio/libs/libxml2.so.2")
library(DESeq2)
library(dplyr)
library(tibble)
library(stringr)
library(apeglm)
library(ggplot2)
dyn.load("/home/rstudio/libs/libgsl.so.25")
dyn.load("/home/rstudio/libs/libgslcblas.so.0")
```

## Init

Set FDR:

```{r, include=T}
fdr = 0.05
```

Define a general function for the differential expression analysis:

```{r, include=T}
diff_expression = function(p, p.cond, diff.levels, de.results.rds.filename) {
  p = round(p)
  
  row.names(p.cond) = names(p)
  
  p.cond$condition = factor(p.cond$condition, levels = diff.levels)
  
  p.dds = DESeqDataSetFromMatrix(countData = p,
                                 colData = p.cond,
                                 design = ~ condition)
  
  p.dds = DESeq(p.dds)
  
  p.res = results(p.dds, 
                  contrast = c("condition", diff.levels[1], diff.levels[2]), 
                  alpha = fdr)
  
  saveRDS(p.res, file = paste0("../r_results/diff_expression/tables/", de.results.rds.filename))
  
  de.results = list(p.res, p.dds)
  
  names(de.results) = c("results_object", "deseq_object")
  
  return(de.results)
}
```

Define a general function to gather differential expression stats:

```{r, include=T}
print_de_summary = function(p.res) {
  cat("Number of significantly differentially regulated genes (FDR =", fdr * 100, "%):", 
    sum(p.res$padj < fdr, na.rm = T), 
    "\n")

  p.n.up = nrow(as.data.frame(p.res) %>% 
    filter(!is.na(padj)) %>%
    filter(padj < fdr) %>%
    filter(log2FoldChange > 0))
  
  cat("Number of upregulated genes (FDR =", fdr * 100, "%):", 
      p.n.up,
      "\n")
  
  p.n.down = nrow(as.data.frame(p.res) %>% 
    filter(!is.na(padj)) %>%
    filter(padj < fdr) %>%
    filter(log2FoldChange < 0))
  
  cat("Number of downregulated genes (FDR =", fdr * 100, "%):", 
      p.n.down,
      "\n")
  
  p.n.all = nrow(as.data.frame(p.res))
  
  cat("The total number of genes:",
      p.n.all,
      "\n")
  
  cat("The percentage of significantly differentially regulated genes:",
      round((p.n.up + p.n.down) / p.n.all * 100, 2))
  
  deg.proportions = list(p.n.up, p.n.down, p.n.all)
  
  names(deg.proportions) = c("up", "down", "all")
  
  return(deg.proportions)
}
```

Define a general function to do PCA:

```{r, include=T}
deg_pca = function(p.dds, pca.plot.filename) {
  p.rld = rlog(p.dds)
  
  # Based on the code from DESeq2 vignette: https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#principal-component-plot-of-the-samples
  p.pcaData = plotPCA(p.rld, returnData = T)
  
  p.percentVar = round(100 * attr(p.pcaData, "percentVar"), 2)
  
  pca.plot = ggplot(p.pcaData, aes(PC1, PC2, color = condition)) +
    geom_point(size = 2) +
    xlab(paste0("PC1: ", p.percentVar[1], "% variance")) +
    ylab(paste0("PC2: ", p.percentVar[2], "% variance")) + 
    coord_fixed() +
    theme_classic()
  
  ggsave(filename = paste0("../r_results/diff_expression/plots/", pca.plot.filename),
         plot = pca.plot)
  
  return(pca.plot)
}
```

Define a general function for plotting the proportions of differentially expressed genes:

```{r, include=T}
plot_deg_proportions = function(p.deseq.object, deg.proportions.plot.filename) {
  p.n.up = p.deseq.object[["up"]]

  p.n.down = p.deseq.object[["down"]]
  
  p.n.all = p.deseq.object[["all"]]

  p.freq.up = p.n.up / p.n.all
  
  p.freq.down = p.n.down / p.n.all
  
  p.freq.nonsign = 1 - p.freq.up - p.freq.down
  
  p.freq.df = data.frame(freq     = c(p.freq.up, p.freq.down, p.freq.nonsign),
                         category = c("Up",      "Down",      "Non-significant"),
                         element_type = "Genes") %>%
    dplyr::mutate(category = factor(category, levels = category))
  
  prop.plot = ggplot(p.freq.df) +
    geom_col(aes(x = element_type,
                 y = freq,
                 fill = category)) + 
    theme_classic()
  
  ggsave(filename = paste0("../r_results/diff_expression/plots/", deg.proportions.plot.filename),
         plot = prop.plot)
  
  return(prop.plot)
}
```

Load the master table:

```{r, include=T}
master.table = assays(readRDS("../input/salmon.merged.gene_counts.rds"))$counts
```

## Generate the reference set of genes in p1 (direct and indirect targets)

Find differentially expressed genes in p1 in D11-MUT vs D11-WT:

```{r, include=T}
p1.raw = master.table %>%
  dplyr::select(MUT_D11_p1_NFIAn_R2,
                MUT_D11_p1_NFIAn_R3,
                WT_D11_p1_NFIAp_R2,
                WT_D11_p1_NFIAp_R3)

# A command like 
# unlist(stringr::str_match_all(names(master.table), "MUT_D11_p1.*")) 
# is convenient for extracting sample names for a given condition 
# from the header of the master table.

p1.cond.raw = data.frame(condition = stringr::str_extract(names(p1.raw), "(MUT|WT)"))

diff.levels = c("MUT", "WT")

p1.results = diff_expression(p1.raw, p1.cond.raw, diff.levels, "deseq-results_p1-MUT-D11-NFIAn_vs_p1-WT-D11-NFIAp.rds")

p1.res = p1.results[["results_object"]]
```

Print a summary of the results:

```{r, include=T}
p1.deg_proportions = print_de_summary(p1.res)
```

Do PCA:

```{r, include=T}
p1.pca = deg_pca(p1.results[["deseq_object"]], "pca_diff-expression_p1-MUT-D11-NFIAn_vs_p1-WT-D11-NFIAp.pdf")

# p1.pca
```

The main difference is not between the conditions, but between on WT sample and the other three samples, which explains the minute number of differentially expressed genes.

## Generate the reference set of genes in p2 (direct and indirect targets)

Find differentially expressed genes in p2 in D11-MUT vs D11-WT:

```{r, include=T}
p2.raw = master.table %>%
  dplyr::select(MUT_D11_p2_NFIAn_R2,
                MUT_D11_p2_NFIAn_R3,
                WT_D11_p2_NFIAp_R2,
                WT_D11_p2_NFIAp_R3)

p2.cond.raw = data.frame(condition = stringr::str_extract(names(p2.raw), "(MUT|WT)"))

diff.levels = c("MUT", "WT")

p2.results = diff_expression(p2.raw, p2.cond.raw, diff.levels, "deseq-results_p2-MUT-D11-NFIAn_vs_p2-WT-D11-NFIAp.rds")

p2.res = p2.results[["results_object"]]
```

Print a summary of the results:

```{r, include=T}
p2.deg_proportions = print_de_summary(p2.res)
```

Do PCA:

```{r, include=T}
p2.pca = deg_pca(p2.results[["deseq_object"]], "pca_diff-expression_p2-MUT-D11-NFIAn_vs_p2-WT-D11-NFIAp.pdf")

# p2.pca
```

The conditions are separated by PC2, but it explains only 22% of the variance, and the samples from the same condition are still quite far apart; hence, the minute number of differentially expressed genes.

## Generate the reference set of genes in pMN (direct and indirect targets)

Find differentially expressed genes in pMN in D11-MUT vs D11-WT:

```{r, include=T}
pM.raw = master.table %>%
  dplyr::select(MUT_D11_pM_NFIAn_R2,
                MUT_D11_pM_NFIAn_R3,
                WT_D11_pM_NFIAp_R2,
                WT_D11_pM_NFIAp_R3)

pM.cond.raw = data.frame(condition = stringr::str_extract(names(pM.raw), "(MUT|WT)"))

diff.levels = c("MUT", "WT")

pM.results = diff_expression(pM.raw, pM.cond.raw, diff.levels, "deseq-results_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp.rds")

pM.res = pM.results[["results_object"]]
```

Print a summary of the results:

```{r, include=T}
pM.deg_proportions = print_de_summary(pM.res)
```

Do PCA:

```{r, include=T}
pM.pca = deg_pca(pM.results[["deseq_object"]], "pca_diff-expression_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp.pdf")

# pM.pca
```

The conditions are separated by PC1, and it explains 71% of the variance; however, the samples from the same condition are still quite far apart from each other. Hence, although there are more differentially expressed genes in pMNs, than in p1 or p2, but still not many.

As there are almost no differentially expressed genes between p{1,2,M}-MUT-D11-NFIAn and p{1,2,M}-WT-D11-NFIAp, while we would expect a considerable number, do some additional checks, taking pMN as an example.

## Check low-expressed genes in all comparisons

Use the `summary()` function from DESeq2 to see low-count genes. If there are a lot of them, then the libraries may be contaminated (for example, with the ribosomal RNA) and so not represent the true RNA diversity; hence, only a low number of differentially expressed genes.

1) Summary of MUT-D11 vs WT-D11 in p1:

```{r, include=T}
summary(p1.res)
```

2) Summary of MUT-D11 vs WT-D11 in p2:

```{r, include=T}
summary(p2.res)
```

3) Summary of MUT-D11 vs WT-D11 in pMN:

```{r, include=T}
summary(pM.res)
```

All the three percentages of low-expressed genes are fine (not really high). So, I cannot say anything about a possible contamination from this.

## Vary parameters of the differential gene expression analysis

1) Show that rising FDR to 10% does not change the number of differential genes dramatically:

```{r, include=T}
fdr = 0.1

pM.results.10 = diff_expression(pM.raw, pM.cond.raw, diff.levels, "deseq-results_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp_fdr10pc.rds")

pM.res.10 = pM.results[["results_object"]]

pM.deg_proportions.10 = print_de_summary(pM.res.10)

fdr = 0.05
```

249 differentially expressed genes, instead of 183 before; hence, not a big difference.

2) Show that the use of the scaled counts does not make any difference:

a) Use `salmon.merged.gene_counts_scaled.rds`:

Load the master table:

```{r, include=T}
master.table.scaled = assays(readRDS("../input/salmon.merged.gene_counts_scaled.rds"))$counts
```

Find differentially expressed genes in pMN in D11-MUT vs D11-WT:

```{r, include=T}
pM.raw.scaled = master.table.scaled %>%
  dplyr::select(MUT_D11_pM_NFIAn_R2,
                MUT_D11_pM_NFIAn_R3,
                WT_D11_pM_NFIAp_R2,
                WT_D11_pM_NFIAp_R3)

pM.cond.raw = data.frame(condition = stringr::str_extract(names(pM.raw.scaled), "(MUT|WT)"))

diff.levels = c("MUT", "WT")

pM.results.scaled = diff_expression(pM.raw.scaled, pM.cond.raw, diff.levels, "deseq-results_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp_scaled.rds")

pM.res.scaled = pM.results.scaled[["results_object"]]

pM.deg_proportions.scaled = print_de_summary(pM.res.scaled)

pM.pca.scaled = deg_pca(pM.results.scaled[["deseq_object"]], "pca_diff-expression_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp_scaled.pdf")

pM.pca.scaled
```

165 differentially expressed genes, instead of 183; even a smaller set.

b) Use `salmon.merged.gene_counts_length_scaled.rds`:

Load the master table:

```{r, include=T}
master.table.scaled = assays(readRDS("../input/salmon.merged.gene_counts_length_scaled.rds"))$counts
```

Find differentially expressed genes in pMN in D11-MUT vs D11-WT:

```{r, include=T}
pM.raw.scaled = master.table.scaled %>%
  dplyr::select(MUT_D11_pM_NFIAn_R2,
                MUT_D11_pM_NFIAn_R3,
                WT_D11_pM_NFIAp_R2,
                WT_D11_pM_NFIAp_R3)

pM.cond.raw = data.frame(condition = stringr::str_extract(names(pM.raw.scaled), "(MUT|WT)"))

diff.levels = c("MUT", "WT")

pM.results.scaled = diff_expression(pM.raw.scaled, pM.cond.raw, diff.levels, "deseq-results_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp_length_scaled.rds")

pM.res.scaled = pM.results.scaled[["results_object"]]

pM.deg_proportions.scaled = print_de_summary(pM.res.scaled)

pM.pca.scaled = deg_pca(pM.results.scaled[["deseq_object"]], "pca_diff-expression_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp_length_scaled.pdf")

pM.pca.scaled
```

183 differentially expressed genes again.

3) Show that pre-filtering of genes by max(expression) >= 10 TPM (with the independent filtering in DESeq2 turned off) does not make any difference:

```{r, include=T}
max_count_cutoff = 10

pM.raw = master.table %>%
  dplyr::select(MUT_D11_pM_NFIAn_R2,
                MUT_D11_pM_NFIAn_R3,
                WT_D11_pM_NFIAp_R2,
                WT_D11_pM_NFIAp_R3)

pM.sample.names = names(pM.raw)

pM.raw = pM.raw %>%
  tibble::rownames_to_column(var = "gene_name") %>%
  rowwise() %>%
  mutate(max_count = max(c_across(all_of(pM.sample.names)))) %>%
  ungroup() %>%
  filter(max_count >= max_count_cutoff) %>%
  as.data.frame() %>%
  dplyr::select(-max_count) %>%
  tibble::column_to_rownames(var = "gene_name")

pM.cond = data.frame(condition = stringr::str_extract(names(pM.raw), "(MUT|WT)"))

diff.levels = c("MUT", "WT")

pM = round(pM.raw)
  
row.names(pM.cond) = names(pM)

pM.cond$condition = factor(pM.cond$condition, levels = diff.levels)

pM.dds = DESeqDataSetFromMatrix(countData = pM,
                                colData = pM.cond,
                                design = ~ condition)

pM.dds = DESeq(pM.dds)

pM.res = results(pM.dds, 
                 contrast = c("condition", diff.levels[1], diff.levels[2]), 
                 alpha = fdr,
                 independentFiltering = F)

saveRDS(pM.res, file = paste0("../r_results/diff_expression/tables/",
                              "deseq-results_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp_length_expr-gte10tpm.rds"))

pM.deg_proportions = print_de_summary(pM.res)

pM.pca = deg_pca(pM.dds, "pca_diff-expression_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp_expr-gte10tpm.pdf")

# pM.pca
```

172 differentially expressed genes, instead of 183; hence, not a big difference. The smaller number of differentially expressed genes obtained after the pre-filtering by max TPM means that the independent filtering by DESeq2 is more efficient (in terms of the number of rejected null hypotheses) than max TPM pre-filtering.

## Check the regulation of the genes from Isabel's poster

```{r, include=t}
control.genes = c("Apcdd1", "Aldh1l1", "Fos", "Nfia")
```

Here I show gene expression counts normalized by DESeq2 (https://hbctraining.github.io/DGE_workshop_salmon/lessons/02_DGE_count_normalization.html), that is, the counts that are used by DESeq2 to perform the differential gene expression analysis.

1) Regulation of Apcdd1, Aldh1l1, Fos and Nfia in p1 between MUT-D11 and WT-D11:

```{r, include=T}
p1.dds = p1.results[["deseq_object"]]

p1.dds = estimateSizeFactors(p1.dds)

p1.norm.counts = counts(p1.dds, normalized = T)

p1.norm.counts %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_name") %>%
  filter(gene_name %in% control.genes) %>%
  left_join(as.data.frame(p1.res) %>%
              rownames_to_column(var = "gene_name"),
            by = c("gene_name" = "gene_name")) %>%
  dplyr::select(-baseMean,
                -lfcSE,
                -stat,
                -pvalue)
```

Aldh1l1, Fos and Nfia are down-regulated, as expected. Apcdd1 is insignificantly upregulated, which does not contradict the graph on the poster. Surprisingly, Nfia is not significantly downregulated, which may be explained by more than a two-fold difference in the DESeq2-normalized counts of this gene between the two replicates of WT-D11.

2) Regulation of Apcdd1, Aldh1l1, Fos and Nfia in p2 between MUT-D11 and WT-D11:

```{r, include=T}
p2.dds = p2.results[["deseq_object"]]

p2.dds = estimateSizeFactors(p2.dds)

p2.norm.counts = counts(p2.dds, normalized = T)

p2.norm.counts %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_name") %>%
  filter(gene_name %in% control.genes) %>%
  left_join(as.data.frame(p2.res) %>%
              rownames_to_column(var = "gene_name"),
            by = c("gene_name" = "gene_name")) %>%
  dplyr::select(-baseMean,
                -lfcSE,
                -stat,
                -pvalue)
```

Strikingly, Nfia is not downregulated, and is not significantly regulated, according to DESeq2. This may highlight a problem with the KO. The other three genes are downregulated, as expected, and even significantly so. But the fact that Nfia is not significantly downregulated (and is not downregulated at all) means that the genes that are differentially expressed between MUT-D11 and WT-D11 in p2 are likely perturbed not because of the (insignificant) change in the Nfia expression, but because of some other reasons. Consequently, I need to check, if this set of differentially expressed genes intersects with the other two sets (in p1 and pMN), as the intersection may be a group of genes whose expression does not actually depend on the Nfia expression.

3) Regulation of Apcdd1, Aldh1l1, Fos and Nfia in pMN between MUT-D11 and WT-D11:

```{r, include=T}
pM.raw = master.table %>%
  dplyr::select(MUT_D11_pM_NFIAn_R2,
                MUT_D11_pM_NFIAn_R3,
                WT_D11_pM_NFIAp_R2,
                WT_D11_pM_NFIAp_R3)

pM.cond.raw = data.frame(condition = stringr::str_extract(names(pM.raw), "(MUT|WT)"))

diff.levels = c("MUT", "WT")

pM.results = diff_expression(pM.raw, pM.cond.raw, diff.levels, "deseq-results_pM-MUT-D11-NFIAn_vs_pM-WT-D11-NFIAp.rds")

pM.dds = pM.results[["deseq_object"]]

pM.res = pM.results[["results_object"]]

pM.dds = estimateSizeFactors(pM.dds)

pM.norm.counts = counts(pM.dds, normalized = T)

pM.norm.counts %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_name") %>%
  filter(gene_name %in% control.genes) %>%
  left_join(as.data.frame(pM.res) %>%
              rownames_to_column(var = "gene_name"),
            by = c("gene_name" = "gene_name")) %>%
  dplyr::select(-baseMean,
                -lfcSE,
                -stat,
                -pvalue)
```

Same here! Seems to be a problem with the Nfia KO. And hence, the even greater number of differentially expressed genes in MUT-D11 vs WT-D11 could be produced not by the changes in the Nfia expression.

## Check NFIA targets among the differentially expressed genes

1) Check the intersection between the genes differentially expressed in p1 and in p2/pMN:

```{r, include=T}
p1.degs = p1.res %>%
  as.data.frame() %>%
  filter(!is.na(padj)) %>%
  filter(padj < fdr) %>%
  rownames_to_column(var = "gene_name") %>%
  pull(gene_name)

p2.degs = p2.res %>%
  as.data.frame() %>%
  filter(!is.na(padj)) %>%
  filter(padj < fdr) %>%
  rownames_to_column(var = "gene_name") %>%
  pull(gene_name)

pM.degs = pM.res %>%
  as.data.frame() %>%
  filter(!is.na(padj)) %>%
  filter(padj < fdr) %>%
  rownames_to_column(var = "gene_name") %>%
  pull(gene_name)

cat("Number of DEGs in p1,  MUT-D11 vs WT-D11          :", length(p1.degs), "\n")
cat("Number of DEGs in p2,  MUT-D11 vs WT-D11          :", length(p2.degs), "\n")
cat("Number of DEGs in pMN, MUT-D11 vs WT-D11          :", length(pM.degs), "\n")
cat("Number of DEGs in common between p2 and p1        :", length(intersect(p1.degs, p2.degs)), "\n")
cat("Number of DEGs in common between pMN and p1       :", length(intersect(p1.degs, pM.degs)), "\n")
cat("Number of DEGs in common between (pMN + p2) and p1:", length(intersect(p1.degs, union(pM.degs, p2.degs))), "\n")
cat("Number of DEGs in common between pMN and p2       :", length(intersect(p2.degs, pM.degs)), "\n")
```

Only 25 DEGs are shared between the union of the p2 and pMN DEGs, where Nfia is not downregulated and hence we do not expect DEGs to depend on Nfia, and p1 DEGs, where Nfia is significantly downregulated. Consequently, I suppose that the majority of 76 DEGs in p1 are indeed the direct or indirect targets of Nfia, and so the set for p1, although much smaller than the other two sets, could be more reliable.

2) Check known NFIA targets among the genes differentially expressed in p1, p2 and pMN:

...

## Generate an alternative reference set of genes in p1 (direct and indirect targets)

As the number of differentially expressed genes in all the three domains between MUT-D11 and WT-D11 is very small, try a differential expression analysis at D9, when NFIA starts to be expressed. If it shows more differentially expressed genes (and if NFIA is significantly downregulated at D9!), I may use the results as a reference set, also taking into account that these results should be enriched in the direct NFIA targets.

There are no NFIAp samples for D9-WT in p1:

```{r, include=T}
unlist(stringr::str_match_all(names(master.table), "WT_D9_p1.*"))
```

so I cannot compare NFIA-expressing p1 cells with the mutant ones.

Additionally, there is only one replicate for MUT-D9 in p2:

```{r, include=T}
unlist(stringr::str_match_all(names(master.table), "MUT_D9_p2.*"))
```

which precludes the differential gene expression analysis in MUT-D9 vs WT-D9 in p2.

However, in pMN there are both more than one replicate of WT-D9-NFIAp and of MUT-D9-NFIAn:

```{r, include=T}
unlist(stringr::str_match_all(names(master.table), "MUT_D9_pM.*"))
unlist(stringr::str_match_all(names(master.table), "WT_D9_pM_NFIAp.*"))
```

hence, I will do the differential gene expression analysis at D9 only for pMN.

Find differentially expressed genes in pMN in D9-MUT vs D9-WT:

```{r, include=T}
pM.raw = master.table %>%
  dplyr::select(MUT_D9_pM_NFIAn_R1,
                MUT_D9_pM_NFIAn_R3,
                WT_D9_pM_NFIAp_R1,
                WT_D9_pM_NFIAp_R2,
                WT_D9_pM_NFIAp_R3)

pM.cond.raw = data.frame(condition = stringr::str_extract(names(pM.raw), "(MUT|WT)"))

diff.levels = c("MUT", "WT")

pM.results = diff_expression(pM.raw, pM.cond.raw, diff.levels, "deseq-results_pM-MUT-D9-NFIAn_vs_pM-WT-D9-NFIAp.rds")

pM.res = pM.results[["results_object"]]
```

Print a summary of the results:

```{r, include=T}
pM.deg_proportions = print_de_summary(pM.res)
```

Same low number of differentially expressed genes.

Do PCA:

```{r, include=T}
pM.pca = deg_pca(pM.results[["deseq_object"]], "pca_diff-expression_pM-MUT-D9-NFIAn_vs_pM-WT-D9-NFIAp.pdf")

# p1.pca
```

The two conditions are separated by PC2 which explains 32% of the variation.

Check the poster genes and Nfia in MUT-D9 vs WT-D9 in pMN:

```{r, include=T}
pM.dds = pM.results[["deseq_object"]]

pM.dds = estimateSizeFactors(pM.dds)

pM.norm.counts = counts(pM.dds, normalized = T)

pM.norm.counts %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_name") %>%
  filter(gene_name %in% control.genes) %>%
  left_join(as.data.frame(pM.res) %>%
              rownames_to_column(var = "gene_name"),
            by = c("gene_name" = "gene_name")) %>%
  dplyr::select(-baseMean,
                -lfcSE,
                -stat,
                -pvalue)
```

Nfia is downregulated, but insignificantly, likely because of the huge difference in expression between WT-R1 vs WT-R2 and WT-R3. Fos and Aldh1l1 are downregulated but insignificantly, while Apcdd1 is not downregulated and is not significantly regulated. According to the poster, we would not expect Apcdd1 and Aldh1l1 to be regulated in either direction, while Fos indeed may be downregulated. But the main conclusion is that there could be a problem with WT replicates and/or with the Nfia KO (Nfia is likely downregulated just because of the very high expression in WT-R1). 

Check the low-expressed genes:

```{r, include=T}
summary(pM.res)
```

17% does not seem to be too high.


