---
title: "predict_correlated_deg"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
dyn.load("/home/rstudio/libs/libxml2.so.2")
library(DESeq2)
library(dplyr)
library(GenomicRanges)
library(tibble)
library(rtracklayer)
library(ggplot2)
library(stringr)
```

Define a function for DESeq2 count normalisation across replicates:

```{r, include=T}
normalize_counts = function(raw.counts, cond.table, diff.levels) {
  raw.counts = round(raw.counts)
  
  row.names(cond.table) = names(raw.counts)
  
  cond.table$condition = factor(cond.table$condition, levels = diff.levels)
  
  dds = DESeqDataSetFromMatrix(countData = raw.counts,
                               colData = cond.table,
                               design = ~ condition)
  
  dds = DESeq(dds)

  dds = estimateSizeFactors(dds)

  return(counts(dds, normalized = T))
}
```

Define a function to generate TSS coordinates from transcript coordinates:

```{r, include=T}
generate_tss = function(tx.ranges, tss.bed.filename) {
  tss.ranges = GRanges(seqnames = seqnames(tx.ranges),
                       ranges = IRanges(start = ifelse(strand(tx.ranges) == "+", 
                                                       start(ranges(tx.ranges)),
                                                       end(ranges(tx.ranges)) - 1),
                                        end = ifelse(strand(tx.ranges) == "+", 
                                                     start(ranges(tx.ranges)) + 1,
                                                     end(ranges(tx.ranges))),
                                        names = names(ranges(tx.ranges))),
                       strand = strand(tx.ranges),
                       gene_name = tx.ranges$gene_name)
  
  export(tss.ranges, tss.bed.filename)
  
  return(tss.ranges)
}
```

Define a function to generate the TSS annotation of the NFIA-dependent genes:

```{r, include=T}
generate_tss_annot = function(p.dep.genes, p.dep.tss.filename) {
  p.dep.genes.annot = mm10.annot.genes[mm10.annot.genes$gene_name %in% p.dep.genes, ]
  
  p.dep.genes.ranges = GRanges(seqnames = p.dep.genes.annot$chr,
                               ranges = IRanges(start = p.dep.genes.annot$start,
                                                end   = p.dep.genes.annot$stop,
                                                names = p.dep.genes.annot$tx_name),
                               strand = strand(p.dep.genes.annot$strand),
                               gene_name = p.dep.genes.annot$gene_name)
  
  p.dep.tss.ranges = generate_tss(p.dep.genes.ranges, p.dep.tss.filename)
}
```

Load the master table:

```{r, include=T}
master.table = assays(readRDS("../input/salmon.merged.gene_counts.rds"))$counts
```

Normalise gene expression across replicates using DESeq2:

```{r, include=T}
# p1

p1.raw.counts = master.table %>%
  dplyr::select(all_of(unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p1.*"))))

p1.cond = data.frame(condition = stringr::str_extract(names(p1.raw.counts), "D(7|9|11)"))

p1.diff.levels = c("D7", "D9", "D11")

p1.norm.counts = normalize_counts(p1.raw.counts, p1.cond, p1.diff.levels)

saveRDS(p1.norm.counts,
        file = "../r_results/predict_correlated_deg/p1_norm_counts_expression.rds")

# p2

p2.raw.counts = master.table %>%
  dplyr::select(all_of(unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p2.*"))))

p2.cond = data.frame(condition = stringr::str_extract(names(p2.raw.counts), "D(7|9|11)"))

p2.diff.levels = c("D7", "D9", "D11")

p2.norm.counts = normalize_counts(p2.raw.counts, p2.cond, p2.diff.levels)

saveRDS(p2.norm.counts,
        file = "../r_results/predict_correlated_deg/p2_norm_counts_expression.rds")

# pMN

pM.raw.counts = master.table %>%
  dplyr::select(all_of(unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_pM.*"))))

pM.cond = data.frame(condition = stringr::str_extract(names(pM.raw.counts), "D(7|9|11)"))

pM.diff.levels = c("D7", "D9", "D11")

pM.norm.counts = normalize_counts(pM.raw.counts, pM.cond, pM.diff.levels)

saveRDS(pM.norm.counts,
        file = "../r_results/predict_correlated_deg/pM_norm_counts_expression.rds")
```

Remove the gene expression master table from memory:

```{r, include=T}
rm(master.table)
```

Upload the chromatin accessibility master table:

```{r, include=T}
master.table = read.delim(file = "../input/consensus_peaks.mRp.clN.featureCounts.txt",
                          header = T,
                          sep = "\t",
                          skip = 1)

names(master.table) = gsub(pattern = ".mLb.clN.bam", replacement = "", x = names(master.table))
```

Normalise chromatin accessibility across replicates using DESeq2:

```{r, include=T}
# p1

p1.raw.counts = master.table %>%
  dplyr::select(all_of(unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p1.*"))))

p1.cond = data.frame(condition = stringr::str_extract(names(p1.raw.counts), "D(7|9|11)"))

p1.diff.levels = c("D7", "D9", "D11")

p1.norm.counts = normalize_counts(p1.raw.counts, p1.cond, p1.diff.levels)

saveRDS(p1.norm.counts,
        file = "../r_results/predict_correlated_deg/p1_norm_counts_accessibility.rds")

# p2

p2.raw.counts = master.table %>%
  dplyr::select(all_of(unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p2.*"))))

p2.cond = data.frame(condition = stringr::str_extract(names(p2.raw.counts), "D(7|9|11)"))

p2.diff.levels = c("D7", "D9", "D11")

p2.norm.counts = normalize_counts(p2.raw.counts, p2.cond, p2.diff.levels)

saveRDS(p2.norm.counts,
        file = "../r_results/predict_correlated_deg/p2_norm_counts_accessibility.rds")

# pMN

pM.raw.counts = master.table %>%
  dplyr::select(all_of(unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_pM.*"))))

pM.cond = data.frame(condition = stringr::str_extract(names(pM.raw.counts), "D(7|9|11)"))

pM.diff.levels = c("D7", "D9", "D11")

pM.norm.counts = normalize_counts(pM.raw.counts, pM.cond, pM.diff.levels)

saveRDS(pM.norm.counts,
        file = "../r_results/predict_correlated_deg/pM_norm_counts_accessibility.rds")
```

Remove the chromatin accessibility master table from memory:

```{r, include=T}
rm(master.table)
```

Upload sets of NFIA-dependent regions:

```{r, include=T}
# Joaquina's parameters for choosing significantly regulated regions
fdr = 0.01
min.l2fc = 2 # Min log2(fold change)
min.baseMean = 100

p1.dep.regions = import(paste0("../r_results/select_diff_regions/p1_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))

p2.dep.regions = import(paste0("../r_results/select_diff_regions/p2_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))

pM.dep.regions = import(paste0("../r_results/select_diff_regions/pM_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))
```

Upload the genome annotation for mm10 that was used for the expression quantification analysis:

```{r, include=T}
# These are actually transcripts, not genes, but I retained the name "genes" 
# to track the provenance of this BED from the original gene.bed used in 
# expression quantification with the nf-core rnaseq pipeline (gene.bed also 
# contains transcripts, not genes).
mm10.tx = read.delim(file = "../results/genes_bed6.bed",
                     header = F)

names(mm10.tx) = c("chr", "start", "stop", "tx_name", "score", "strand")

mm10.gene2tx = read.delim(file = "../results/mm10_genes_to_transcripts.tsv",
                          header = F)

names(mm10.gene2tx) = c("gene_name", "tx_name")

mm10.annot.genes = mm10.tx %>%
  left_join(mm10.gene2tx,
            by = c("tx_name" = "tx_name"))
```

Upload sets of NFIA-dependent genes:

```{r, include=T}
p1.dep.genes = readRDS(file = "../r_results/diff_expression/tables/p1_ref_genes.rds")

p2.dep.genes = readRDS(file = "../r_results/diff_expression/tables/p2_ref_genes.rds")

pM.dep.genes = readRDS(file = "../r_results/diff_expression/tables/pM_ref_genes.rds")
```

Check that all NFIA-dependent genes are present in the genome annotation:

```{r, include=T}
cat("Number of p1 NFIA-dependent genes absent from the annotation:", length(p1.dep.genes) - sum(p1.dep.genes %in% mm10.annot.genes$gene_name), "\n")

cat("Number of p2 NFIA-dependent genes absent from the annotation:", length(p2.dep.genes) - sum(p2.dep.genes %in% mm10.annot.genes$gene_name), "\n")

cat("Number of pM NFIA-dependent genes absent from the annotation:", length(pM.dep.genes) - sum(pM.dep.genes %in% mm10.annot.genes$gene_name), "\n")
```

Generate genomic ranges for the NFIA-dependent genes:

```{r, include=T}
p1.dep.tss.ranges = generate_tss_annot(p1.dep.genes,
                                       "../r_results/predict_correlated_deg/p1_dep_genes_tss.bed")

p2.dep.tss.ranges = generate_tss_annot(p2.dep.genes,
                                       "../r_results/predict_correlated_deg/p2_dep_genes_tss.bed")

pM.dep.tss.ranges = generate_tss_annot(pM.dep.genes,
                                       "../r_results/predict_correlated_deg/pM_dep_genes_tss.bed")
```



