---
title: "predict_correlated_deg"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
dyn.load("/home/rstudio/libs/libxml2.so.2")
library(DESeq2)
library(dplyr)
library(GenomicRanges)
library(tibble)
library(rtracklayer)
library(ggplot2)
library(stringr)
library(scales)
library(tidyr)
```

Load helper functions:

```{r, include=T}
source("helper_functions.R")
```

Define constants:

```{r, include=T}
set.seed(10) # reproduce exactly the same sequence of pseudorandom numbers each run

# Parameters for choosing significantly regulated regions
fdr = 0.01
min.l2fc = 2 # Min log2(fold change) | 1
min.baseMean = 100 # 20, 40

fdr.matches = 0.05

vicinity.radius = 500 # kbp

pcc.fdr = 0.05

expression.median = 5 # minimal median expression of a gene across all samples to count the gene as expressed

output.suffix = paste0("_expr-median", expression.median)

genes.to.exclude = "ERCC-" # exclude spike-in control
```

Load the master table:

```{r, include=T}
master.table = assays(readRDS("../input/salmon.merged.gene_counts.rds"))$counts %>%
  rownames_to_column(var = "gene_name") %>%
  filter(!stringr::str_detect(gene_name, fixed(genes.to.exclude))) %>%
  column_to_rownames(var = "gene_name")

gene.names = rownames(master.table)
```

Normalise gene expression across replicates using DESeq2:

```{r, include=T}
p1.norm.gene.counts.deseq2 = generate_norm_gene_counts(master.table, "p1")

p2.norm.gene.counts.deseq2 = generate_norm_gene_counts(master.table, "p2")

pM.norm.gene.counts.deseq2 = generate_norm_gene_counts(master.table, "pM")
```

Select expressed genes:

```{r, include=T}
p1.expr.genes = select_expressed_genes(p1.norm.gene.counts.deseq2, expression.median)

p2.expr.genes = select_expressed_genes(p2.norm.gene.counts.deseq2, expression.median)

pM.expr.genes = select_expressed_genes(pM.norm.gene.counts.deseq2, expression.median)

cat("Number of expressed genes in p1 ( expression.median =", expression.median, "):", length(p1.expr.genes), "\n")

cat("Number of expressed genes in p2 ( expression.median =", expression.median, "):", length(p2.expr.genes), "\n")

cat("Number of expressed genes in pM ( expression.median =", expression.median, "):", length(pM.expr.genes), "\n")
```

Scale the normalised expression counts (calculate z-scores) of expressed genes:

```{r, include=T}
p1.norm.gene.counts = generate_scaled_counts(p1.norm.gene.counts.deseq2[p1.norm.gene.counts.deseq2$gene_names %in% p1.expr.genes, ], "gene_names")

p2.norm.gene.counts = generate_scaled_counts(p2.norm.gene.counts.deseq2[p2.norm.gene.counts.deseq2$gene_names %in% p2.expr.genes, ], "gene_names")

pM.norm.gene.counts = generate_scaled_counts(pM.norm.gene.counts.deseq2[pM.norm.gene.counts.deseq2$gene_names %in% pM.expr.genes, ], "gene_names")
```

Write down the RNA-seq sample names:

```{r, include=T}
p1.gene.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p1.*"))

p2.gene.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p2.*"))

pM.gene.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_pM.*"))
```

Remove the gene expression master table from memory:

```{r, include=T}
rm(master.table)
```

Upload the chromatin accessibility master table:

```{r, include=T}
master.table = read.delim(file = "../input/consensus_peaks.mRp.clN.featureCounts.txt",
                          header = T,
                          sep = "\t",
                          skip = 1)

names(master.table) = gsub(pattern = ".mLb.clN.bam", replacement = "", x = names(master.table))

region.annot = master.table %>%
  dplyr::select(Geneid,
                Chr,
                Start,
                End,
                Strand,
                Length)
```

Normalise chromatin accessibility across replicates using DESeq2:

```{r, include=T}
p1.norm.region.counts.deseq2 = generate_norm_chrom_counts(master.table, region.annot, "p1")

p2.norm.region.counts.deseq2 = generate_norm_chrom_counts(master.table, region.annot, "p2")

pM.norm.region.counts.deseq2 = generate_norm_chrom_counts(master.table, region.annot, "pM")
```

Scale the normalised region accessibility (calculate z-scores):

```{r, include=T}
p1.norm.region.counts = generate_scaled_counts(p1.norm.region.counts.deseq2, "region_names")

p2.norm.region.counts = generate_scaled_counts(p2.norm.region.counts.deseq2, "region_names")

pM.norm.region.counts = generate_scaled_counts(pM.norm.region.counts.deseq2, "region_names")
```

Write down the ATAC-seq sample names:

```{r, include=T}
p1.region.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p1.*"))

p2.region.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p2.*"))

pM.region.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_pM.*"))
```

Remove the chromatin accessibility master table from memory:

```{r, include=T}
rm(master.table)
```

Upload sets of NFIA-dependent regions:

```{r, include=T}
p1.dep.regions = import(paste0("../r_results/select_diff_regions/p1_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))

p2.dep.regions = import(paste0("../r_results/select_diff_regions/p2_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))

pM.dep.regions = import(paste0("../r_results/select_diff_regions/pM_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))
```

Print the number of NFIA-dependet regions:

```{r, include=T}
cat("Number of NFIA-dependent regions in p1:", length(p1.dep.regions$name), "\n")

cat("Number of NFIA-dependent regions in p2:", length(p2.dep.regions$name), "\n")

cat("Number of NFIA-dependent regions in pM:", length(pM.dep.regions$name), "\n")
```

Upload the genome annotation for mm10 that was used for the expression quantification analysis:

```{r, include=T}
# These are actually transcripts, not genes, but I retained the name "genes" 
# to track the provenance of this BED from the original gene.bed used in 
# expression quantification with the nf-core rnaseq pipeline (gene.bed also 
# contains transcripts, not genes).
mm10.tx = read.delim(file = "../results/genes_bed6.bed",
                     header = F)

names(mm10.tx) = c("chr", "start", "stop", "tx_name", "score", "strand")

mm10.gene2tx = read.delim(file = "../results/mm10_genes_to_transcripts.tsv",
                          header = F)

names(mm10.gene2tx) = c("gene_name", "tx_name")

mm10.annot.genes = mm10.tx %>%
  left_join(mm10.gene2tx,
            by = c("tx_name" = "tx_name"))
```

Generate genomic ranges for the TSSs of the expressed genes:

```{r, include=T}
p1.expr.tss.ranges = generate_tss_annot(p1.expr.genes,
                                        mm10.annot.genes,
                                        "../r_results/predict_correlated_expressed_gene_random/p1_expr_genes_tss.bed")

p2.expr.tss.ranges = generate_tss_annot(p2.expr.genes,
                                        mm10.annot.genes,
                                        "../r_results/predict_correlated_expressed_gene_random/p2_expr_genes_tss.bed")

pM.expr.tss.ranges = generate_tss_annot(pM.expr.genes,
                                        mm10.annot.genes,
                                        "../r_results/predict_correlated_expressed_gene_random/pM_expr_genes_tss.bed")
```

Calculate the distribution of distances between NFIA-dependent regions and the closest expressed TSSs (on the same strand) to make sure that regions are mostly enhancers, as it would make sense to analyse promoters (regions overlapping TSSs) separately:

```{r, include=T}
calc_dist_distributions(p1.dep.regions, p1.expr.tss.ranges, "p1")

calc_dist_distributions(p2.dep.regions, p2.expr.tss.ranges, "p2")

calc_dist_distributions(pM.dep.regions, pM.expr.tss.ranges, "pM")
```

Up to 2% of NFIA-dependent regions are in core or proximal promoters, hence I will need to analyse them separately (and, probably, to assign them to the closest expressed TSS). For now, I continue to analyse all NFIA-dependent regions together, as the proportion of promoter ones is very small in any of the three domains.

Upload mm10 chromosome sizes:

```{r, include=T}
mm10.chr.sizes.df = read.delim(file = "../input/mm10.chrom.sizes",
                               header = F,
                               sep = "\t")

mm10.chr.sizes = mm10.chr.sizes.df$V2

names(mm10.chr.sizes) = mm10.chr.sizes.df$V1
```

Generate the region-gene assignment areas as vicinities of a fixed radius around the NFIA-dependent regions:

```{r, include=T}
p1.vicinity.radius = generate_vicinity_radius(p1.dep.regions, 
                                              vicinity.radius,
                                              paste0("../r_results/predict_correlated_expressed_gene_random/p1_vicinities_radius_", 
                                                     vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc, 
                                                     "_min-baseMean", min.baseMean, ".bed"))

p2.vicinity.radius = generate_vicinity_radius(p2.dep.regions, 
                                              vicinity.radius,
                                              paste0("../r_results/predict_correlated_expressed_gene_random/p2_vicinities_radius_", 
                                                     vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc, 
                                                     "_min-baseMean", min.baseMean, ".bed"))

pM.vicinity.radius = generate_vicinity_radius(pM.dep.regions, 
                                              vicinity.radius,
                                              paste0("../r_results/predict_correlated_expressed_gene_random/pM_vicinities_radius_", 
                                                     vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc, 
                                                     "_min-baseMean", min.baseMean, ".bed"))
```

Calculate Pearson correlation coefficients (PCCs) between NFIA-dependent regions and expressed genes within the assignment areas defined by a fixed radius around the regions:

```{r, include=T}
p1.corr.radius = calc_correlations(p1.vicinity.radius, p1.expr.tss.ranges,
                                   p1.norm.region.counts, p1.norm.gene.counts,
                                   p1.region.sample.names, p1.gene.sample.names,
                                   "p1",
                                   paste0("../r_results/predict_correlated_expressed_gene_random/p1_corr_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))

p2.corr.radius = calc_correlations(p2.vicinity.radius, p2.expr.tss.ranges,
                                   p2.norm.region.counts, p2.norm.gene.counts,
                                   p2.region.sample.names, p2.gene.sample.names,
                                   "p2",
                                   paste0("../r_results/predict_correlated_expressed_gene_random/p2_corr_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))

pM.corr.radius = calc_correlations(pM.vicinity.radius, pM.expr.tss.ranges,
                                   pM.norm.region.counts, pM.norm.gene.counts,
                                   pM.region.sample.names, pM.gene.sample.names,
                                   "pM",
                                   paste0("../r_results/predict_correlated_expressed_gene_random/pM_corr_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))
```

Generate background PCC distributions: For each NFIA-dependent region, take all expressed genes:

```{r, include=T}
p1.bkgd.radius = calc_correlations_pairs_all(p1.dep.regions, 
                                             p1.norm.region.counts, p1.norm.gene.counts,
                                             p1.region.sample.names, p1.gene.sample.names,
                                             "p1",
                                             paste0("../r_results/predict_correlated_expressed_gene_random/p1_bkgd_in_radius_",
                                                    vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                                    "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))

p2.bkgd.radius = calc_correlations_pairs_all(p2.dep.regions, 
                                             p2.norm.region.counts, p2.norm.gene.counts,
                                             p2.region.sample.names, p2.gene.sample.names,
                                             "p2",
                                             paste0("../r_results/predict_correlated_expressed_gene_random/p2_bkgd_in_radius_",
                                                    vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                                    "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))

pM.bkgd.radius = calc_correlations_pairs_all(pM.dep.regions, 
                                             pM.norm.region.counts, pM.norm.gene.counts,
                                             pM.region.sample.names, pM.gene.sample.names,
                                             "pM",
                                             paste0("../r_results/predict_correlated_expressed_gene_random/pM_bkgd_in_radius_",
                                                    vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                                    "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))
```

Plot the background and the target distributions of the absolute PCC values:

```{r, include=T}
p1.corr.bkgd.radius = p1.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(p1.bkgd.radius %>%
              mutate(pair_type = "Background"))
  
p1.plot = p1.corr.bkgd.radius %>%
  ggplot(aes(x = abs_pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/p1_corr_and_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_density.pdf"),
       plot = p1.plot)

p2.corr.bkgd.radius = p2.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(p2.bkgd.radius %>%
              mutate(pair_type = "Background"))
  
p2.plot = p2.corr.bkgd.radius %>%
  ggplot(aes(x = abs_pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/p2_corr_and_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_density.pdf"),
       plot = p2.plot)

pM.corr.bkgd.radius = pM.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(pM.bkgd.radius %>%
              mutate(pair_type = "Background"))
  
pM.plot = pM.corr.bkgd.radius %>%
  ggplot(aes(x = abs_pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/pM_corr_and_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_density.pdf"),
       plot = pM.plot)
```

The target distributions of absolute PCCs are slightly enriched in higher values.

Plot the background and the target distributions of the PCC values:

```{r, include=T}
p1.corr.bkgd.radius = p1.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(p1.bkgd.radius %>%
              mutate(pair_type = "Background"))
  
p1.plot = p1.corr.bkgd.radius %>%
  ggplot(aes(x = pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/p1_corr_and_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_pcc_density.pdf"),
       plot = p1.plot)

p2.corr.bkgd.radius = p2.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(p2.bkgd.radius %>%
              mutate(pair_type = "Background"))
  
p2.plot = p2.corr.bkgd.radius %>%
  ggplot(aes(x = pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/p2_corr_and_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_pcc_density.pdf"),
       plot = p2.plot)

pM.corr.bkgd.radius = pM.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(pM.bkgd.radius %>%
              mutate(pair_type = "Background"))
  
pM.plot = pM.corr.bkgd.radius %>%
  ggplot(aes(x = pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/pM_corr_and_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_pcc_density.pdf"),
       plot = pM.plot)
```

The distributions are not normal, so I do not see any reason for using them instead of the distributions of absolute PCCs.

Calculate empirical p-values of absolute PCCs:

```{r, include=T}
p1.bkgd.df_nrow = nrow(p1.bkgd.radius)

p1.abs_pcc_sorted = p1.bkgd.radius %>%
  arrange(-abs_pcc) %>%
  pull(abs_pcc)

p1.corr.radius.sign = p1.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, p1.abs_pcc_sorted, p1.bkgd.df_nrow)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in p1, radius:", nrow(p1.corr.radius.sign %>% filter(sign.padj)), "\n")

p2.bkgd.df_nrow = nrow(p2.bkgd.radius)

p2.abs_pcc_sorted = p2.bkgd.radius %>%
  arrange(-abs_pcc) %>%
  pull(abs_pcc)

p2.corr.radius.sign = p2.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, p2.abs_pcc_sorted, p2.bkgd.df_nrow)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in p2, radius:", nrow(p2.corr.radius.sign %>% filter(sign.padj)), "\n")

pM.bkgd.df_nrow = nrow(pM.bkgd.radius)

pM.abs_pcc_sorted = pM.bkgd.radius %>%
  arrange(-abs_pcc) %>%
  pull(abs_pcc)

pM.corr.radius.sign = pM.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, pM.abs_pcc_sorted, pM.bkgd.df_nrow)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in pM, radius:", nrow(pM.corr.radius.sign %>% filter(sign.padj)), "\n")
```

No significant pairs. Indeed, NFIA-dependent regions very highly correlate with each other, and we expect regulated genes to be highly correlated with the corresponding NFIA-dependent regions. Therefore, we have to expect that an NFIA-dependent region is as well correlated with regulated genes from other vicinities as with regulated genes from its own vicinity. Consequently, if a large proportion of background region-gene pairs actually contain expressed genes from region vicinities, then the background PCC distribution may be too close to the target one to be a proper background.

First, check the proportions of expressed genes that are within at least one vicinity of NFIA-dependent regions. A gene is within a vicinity if at least one of its TSSs is within this vicinity:

```{r, include=T}
p1.genes.in.vicinity.n = count_expr_genes_inside_vicinity(p1.vicinity.radius, p1.expr.tss.ranges)

p2.genes.in.vicinity.n = count_expr_genes_inside_vicinity(p2.vicinity.radius, p2.expr.tss.ranges)

pM.genes.in.vicinity.n = count_expr_genes_inside_vicinity(pM.vicinity.radius, pM.expr.tss.ranges)

cat("The number of expressed genes in NFIA-dependent region vicinities in p1:", p1.genes.in.vicinity.n, 
    "(", p1.genes.in.vicinity.n / length(p1.expr.genes) * 100,"% )\n")

cat("The number of expressed genes in NFIA-dependent region vicinities in p1:", p2.genes.in.vicinity.n, 
    "(", p2.genes.in.vicinity.n / length(p2.expr.genes) * 100,"% )\n")

cat("The number of expressed genes in NFIA-dependent region vicinities in p1:", pM.genes.in.vicinity.n, 
    "(", pM.genes.in.vicinity.n / length(pM.expr.genes) * 100,"% )\n")
```

From a third to a half of all expressed genes are in NFIA-dependent region vicinities.

Check the proportions of region-gene pairs with expressed genes from vicinities. It should match the proportions of expressed genes above:

```{r, include=T}
p1.genes.in.vicinity.names = find_expr_genes_inside_vicinity(p1.vicinity.radius, p1.expr.tss.ranges)

p2.genes.in.vicinity.names = find_expr_genes_inside_vicinity(p2.vicinity.radius, p2.expr.tss.ranges)

pM.genes.in.vicinity.names = find_expr_genes_inside_vicinity(pM.vicinity.radius, pM.expr.tss.ranges)

p1.bkgd.pairs.with.vicinity.genes = p1.bkgd.radius %>%
  filter(gene_name %in% p1.genes.in.vicinity.names)

p2.bkgd.pairs.with.vicinity.genes = p2.bkgd.radius %>%
  filter(gene_name %in% p2.genes.in.vicinity.names)

pM.bkgd.pairs.with.vicinity.genes = pM.bkgd.radius %>%
  filter(gene_name %in% pM.genes.in.vicinity.names)

cat("The number of background region-gene pairs with vicinity-located genes:", nrow(p1.bkgd.pairs.with.vicinity.genes), 
    "(", nrow(p1.bkgd.pairs.with.vicinity.genes) / nrow(p1.bkgd.radius) * 100, "% )\n")

cat("The number of background region-gene pairs with vicinity-located genes:", nrow(p2.bkgd.pairs.with.vicinity.genes), 
    "(", nrow(p2.bkgd.pairs.with.vicinity.genes) / nrow(p2.bkgd.radius) * 100, "% )\n")

cat("The number of background region-gene pairs with vicinity-located genes:", nrow(pM.bkgd.pairs.with.vicinity.genes), 
    "(", nrow(pM.bkgd.pairs.with.vicinity.genes) / nrow(pM.bkgd.radius) * 100, "% )\n")
```

Correct.

To check if the region-gene pairs with vicinity-located expressed genes give the higher absolute PCCs, stratify the background absolute PCC distribution by the type of an expressed gene in each pair (vicinity-located or not):

```{r, include=T}
p1.plot = p1.bkgd.radius %>%
  left_join(data.frame(gene_name = p1.expr.genes) %>%
              mutate(gene_type = ifelse(gene_name %in% p1.genes.in.vicinity.names,
                                        "Vicinity",
                                        "Non-vicinity")),
            by = c("gene_name" = "gene_name")) %>%
  ggplot(aes(x = abs_pcc,
             fill = gene_type)) +
    geom_density(alpha = 0.5) +
    theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/p1_bkgd_by-gene-type_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_density.pdf"),
       plot = p1.plot)

p2.plot = p2.bkgd.radius %>%
  left_join(data.frame(gene_name = p2.expr.genes) %>%
              mutate(gene_type = ifelse(gene_name %in% p2.genes.in.vicinity.names,
                                        "Vicinity",
                                        "Non-vicinity")),
            by = c("gene_name" = "gene_name")) %>%
  ggplot(aes(x = abs_pcc,
             fill = gene_type)) +
    geom_density(alpha = 0.5) +
    theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/p2_bkgd_by-gene-type_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_density.pdf"),
       plot = p2.plot)

pM.plot = pM.bkgd.radius %>%
  left_join(data.frame(gene_name = pM.expr.genes) %>%
              mutate(gene_type = ifelse(gene_name %in% pM.genes.in.vicinity.names,
                                        "Vicinity",
                                        "Non-vicinity")),
            by = c("gene_name" = "gene_name")) %>%
  ggplot(aes(x = abs_pcc,
             fill = gene_type)) +
    geom_density(alpha = 0.5) +
    theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/pM_bkgd_by-gene-type_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_density.pdf"),
       plot = pM.plot)
```

Check that there are no zero raw p-values in the original results:

```{r, include=T}
p1.corr.radius.sign %>%
  arrange(pvalue) %>%
  filter(pvalue == 0)

p2.corr.radius.sign %>%
  arrange(pvalue) %>%
  filter(pvalue == 0)

pM.corr.radius.sign %>%
  arrange(pvalue) %>%
  filter(pvalue == 0)
```

No zero raw p-values.

Actually, a more correct way of defining the background would be to take only guaranteed non-regulatory region-gene pairs, instead of all pairs, as the background probability distribution should, by definition, correspond to the null hypothesis of a lack of regulation between a region and a gene. Therefore, to build such a distribution, we need to exclude pairs of NFIA-dependent regions with expressed genes on the same chromosome, as any of these genes may potentially be regulated by any of the regions on the chromosome.

Build a modified background without same-chromosome pairs:

```{r, include=T}
p1.bkgd.radius.outside.chr = calc_correlations_pairs_outside_chr(p1.dep.regions, p1.expr.tss.ranges,
                                                                 p1.norm.region.counts, p1.norm.gene.counts,
                                                                 p1.region.sample.names, p1.gene.sample.names,
                                                                 "p1",
                                                                 paste0("../r_results/predict_correlated_expressed_gene_random/p1_bkgd_outside_chr_in_radius_",
                                                                        vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                                                        "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))

p2.bkgd.radius.outside.chr = calc_correlations_pairs_outside_chr(p2.dep.regions, p2.expr.tss.ranges,
                                                                 p2.norm.region.counts, p2.norm.gene.counts,
                                                                 p2.region.sample.names, p2.gene.sample.names,
                                                                 "p2",
                                                                 paste0("../r_results/predict_correlated_expressed_gene_random/p2_bkgd_outside_chr_in_radius_",
                                                                        vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                                                        "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))

pM.bkgd.radius.outside.chr = calc_correlations_pairs_outside_chr(pM.dep.regions, pM.expr.tss.ranges,
                                                                 pM.norm.region.counts, pM.norm.gene.counts,
                                                                 pM.region.sample.names, pM.gene.sample.names,
                                                                 "pM",
                                                                 paste0("../r_results/predict_correlated_expressed_gene_random/pM_bkgd_outside_chr_in_radius_",
                                                                        vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                                                        "_min-baseMean", min.baseMean, "_", output.suffix, ".rds"))
```

Plot the background and the target distributions of the absolute PCC values:

```{r, include=T}
p1.corr.bkgd.radius = p1.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(p1.bkgd.radius.outside.chr %>%
              mutate(pair_type = "Background"))
  
p1.plot = p1.corr.bkgd.radius %>%
  ggplot(aes(x = abs_pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/p1_corr_and_bkgd_outside-chr_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_density.pdf"),
       plot = p1.plot)

p2.corr.bkgd.radius = p2.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(p2.bkgd.radius.outside.chr %>%
              mutate(pair_type = "Background"))
  
p2.plot = p2.corr.bkgd.radius %>%
  ggplot(aes(x = abs_pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/p2_corr_and_bkgd_outside-chr_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_density.pdf"),
       plot = p2.plot)

pM.corr.bkgd.radius = pM.corr.radius %>%
  mutate(pair_type = "Target") %>%
  bind_rows(pM.bkgd.radius.outside.chr %>%
              mutate(pair_type = "Background"))
  
pM.plot = pM.corr.bkgd.radius %>%
  ggplot(aes(x = abs_pcc,
             fill = pair_type)) +
  geom_density(alpha = 0.5) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_expressed_gene_random/plots/pM_corr_and_bkgd_outside-chr_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_", output.suffix, "_density.pdf"),
       plot = pM.plot)
```

No difference with the all-regions-to-all-expressed-genes background.

Calculate empirical p-values of absolute PCCs:

```{r, include=T}
p1.bkgd.df_nrow = nrow(p1.bkgd.radius.outside.chr)

p1.abs_pcc_sorted = p1.bkgd.radius.outside.chr %>%
  arrange(-abs_pcc) %>%
  pull(abs_pcc)

p1.corr.radius.sign = p1.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, p1.abs_pcc_sorted, p1.bkgd.df_nrow)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in p1, radius:", nrow(p1.corr.radius.sign %>% filter(sign.padj)), "\n")

p2.bkgd.df_nrow = nrow(p2.bkgd.radius.outside.chr)

p2.abs_pcc_sorted = p2.bkgd.radius.outside.chr %>%
  arrange(-abs_pcc) %>%
  pull(abs_pcc)

p2.corr.radius.sign = p2.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, p2.abs_pcc_sorted, p2.bkgd.df_nrow)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in p2, radius:", nrow(p2.corr.radius.sign %>% filter(sign.padj)), "\n")

pM.bkgd.df_nrow = nrow(pM.bkgd.radius.outside.chr)

pM.abs_pcc_sorted = pM.bkgd.radius.outside.chr %>%
  arrange(-abs_pcc) %>%
  pull(abs_pcc)

pM.corr.radius.sign = pM.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, pM.abs_pcc_sorted, pM.bkgd.df_nrow)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in pM, radius:", nrow(pM.corr.radius.sign %>% filter(sign.padj)), "\n")
```

No significant region-gene pairs.
