---
title: "predict_correlated_deg"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
dyn.load("/home/rstudio/libs/libxml2.so.2")
library(DESeq2)
library(dplyr)
library(GenomicRanges)
library(tibble)
library(rtracklayer)
library(ggplot2)
library(stringr)
library(scales)
```

Load helper functions:

```{r, include=T}
source("helper_functions.R")
```

Define constants:

```{r, include=T}
fdr.matches = 0.05

expression.cutoff = 5 # min number of normalised reads

pass.count.cutoff = 2 # min number of RNA-seq samples with at least expression.cutoff normalised reads

vicinity.radius = 500 # kbp

pcc.fdr = 0.05
```

Define cutoff sets:

```{r, include=T}
region.cutoffs = list("set1" = list("fdr" = 0.01, "min.l2fc" = 2, "min.baseMean" = 100),
                      "set2" = list("fdr" = 0.01, "min.l2fc" = 2, "min.baseMean" = 80),
                      "set3" = list("fdr" = 0.01, "min.l2fc" = 2, "min.baseMean" = 60),
                      "set4" = list("fdr" = 0.01, "min.l2fc" = 2, "min.baseMean" = 40),
                      "set5" = list("fdr" = 0.01, "min.l2fc" = 2, "min.baseMean" = 20),
                      "set6" = list("fdr" = 0.01, "min.l2fc" = 1.5, "min.baseMean" = 20),
                      "set7" = list("fdr" = 0.01, "min.l2fc" = 1, "min.baseMean" = 20),
                      "set8" = list("fdr" = 0.01, "min.l2fc" = 0.5, "min.baseMean" = 20),
                      "set9" = list("fdr" = 0.02, "min.l2fc" = 0.5, "min.baseMean" = 20),
                      "set10" = list("fdr" = 0.03, "min.l2fc" = 0.5, "min.baseMean" = 20),
                      "set11" = list("fdr" = 0.04, "min.l2fc" = 0.5, "min.baseMean" = 20),
                      "set12" = list("fdr" = 0.05, "min.l2fc" = 0.5, "min.baseMean" = 20))
```

Load the master table:

```{r, include=T}
master.table = assays(readRDS("../input/salmon.merged.gene_counts.rds"))$counts

gene.names = rownames(master.table)
```

Normalise gene expression across replicates using DESeq2:

```{r, include=T}
p1.norm.gene.counts = generate_norm_gene_counts(master.table, "p1")

p2.norm.gene.counts = generate_norm_gene_counts(master.table, "p2")

pM.norm.gene.counts = generate_norm_gene_counts(master.table, "pM")
```

Write down the RNA-seq sample names:

```{r, include=T}
p1.gene.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p1.*"))

p2.gene.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p2.*"))

pM.gene.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_pM.*"))
```

Remove the gene expression master table from memory:

```{r, include=T}
rm(master.table)
```

Upload the chromatin accessibility master table:

```{r, include=T}
master.table = read.delim(file = "../input/consensus_peaks.mRp.clN.featureCounts.txt",
                          header = T,
                          sep = "\t",
                          skip = 1)

names(master.table) = gsub(pattern = ".mLb.clN.bam", replacement = "", x = names(master.table))

region.annot = master.table %>%
  dplyr::select(Geneid,
                Chr,
                Start,
                End,
                Strand,
                Length)
```

Normalise chromatin accessibility across replicates using DESeq2:

```{r, include=T}
p1.norm.region.counts = generate_norm_chrom_counts(master.table, region.annot, "p1")

p2.norm.region.counts = generate_norm_chrom_counts(master.table, region.annot, "p2")

pM.norm.region.counts = generate_norm_chrom_counts(master.table, region.annot, "pM")
```

Write down the ATAC-seq sample names:

```{r, include=T}
p1.region.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p1.*"))

p2.region.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_p2.*"))

pM.region.sample.names = unlist(stringr::str_match_all(names(master.table), "WT_D[179]+_pM.*"))
```

Remove the chromatin accessibility master table from memory:

```{r, include=T}
rm(master.table)
```

Upload sets of NFIA-dependent regions:

```{r, include=T}
# Joaquina's parameters for choosing significantly regulated regions
fdr = 0.01
min.l2fc = 2 # Min log2(fold change)
min.baseMean = 100

p1.dep.regions = import(paste0("../r_results/select_diff_regions/p1_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))

p2.dep.regions = import(paste0("../r_results/select_diff_regions/p2_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))

pM.dep.regions = import(paste0("../r_results/select_diff_regions/pM_dep_ranges", 
                               "_fdr", fdr, "_min-l2fc", min.l2fc, "_min-baseMean", min.baseMean,
                               ".bed"))
```

Upload the genome annotation for mm10 that was used for the expression quantification analysis:

```{r, include=T}
# These are actually transcripts, not genes, but I retained the name "genes" 
# to track the provenance of this BED from the original gene.bed used in 
# expression quantification with the nf-core rnaseq pipeline (gene.bed also 
# contains transcripts, not genes).
mm10.tx = read.delim(file = "../results/genes_bed6.bed",
                     header = F)

names(mm10.tx) = c("chr", "start", "stop", "tx_name", "score", "strand")

mm10.gene2tx = read.delim(file = "../results/mm10_genes_to_transcripts.tsv",
                          header = F)

names(mm10.gene2tx) = c("gene_name", "tx_name")

mm10.annot.genes = mm10.tx %>%
  left_join(mm10.gene2tx,
            by = c("tx_name" = "tx_name"))
```

Select expressed genes:

```{r, include=T}
p1.expr.genes = select_expressed_genes(p1.norm.gene.counts)

p2.expr.genes = select_expressed_genes(p2.norm.gene.counts)

pM.expr.genes = select_expressed_genes(pM.norm.gene.counts)

cat("Number of expressed genes in p1 ( expression.cutoff =", expression.cutoff, "):", length(p1.expr.genes), "\n")

cat("Number of expressed genes in p2 ( expression.cutoff =", expression.cutoff, "):", length(p2.expr.genes), "\n")

cat("Number of expressed genes in pM ( expression.cutoff =", expression.cutoff, "):", length(pM.expr.genes), "\n")
```

Generate genomic ranges for the TSSs of the expressed genes:

```{r, include=T}
p1.expr.tss.ranges = generate_tss_annot(p1.expr.genes,
                                        mm10.annot.genes,
                                        "../r_results/predict_correlated_expressed_gene/p1_expr_genes_tss.bed")

p2.expr.tss.ranges = generate_tss_annot(p2.expr.genes,
                                        mm10.annot.genes,
                                        "../r_results/predict_correlated_expressed_gene/p2_expr_genes_tss.bed")

pM.expr.tss.ranges = generate_tss_annot(pM.expr.genes,
                                        mm10.annot.genes,
                                        "../r_results/predict_correlated_expressed_gene/pM_expr_genes_tss.bed")
```

Upload mm10 chromosome sizes:

```{r, include=T}
mm10.chr.sizes.df = read.delim(file = "../input/mm10.chrom.sizes",
                               header = F,
                               sep = "\t")

mm10.chr.sizes = mm10.chr.sizes.df$V2

names(mm10.chr.sizes) = mm10.chr.sizes.df$V1
```

Generate the region-gene assignment areas as vicinities of a fixed radius around the NFIA-dependent regions:

```{r, include=T}
p1.vicinity.radius = generate_vicinity_radius(p1.dep.regions, 
                                              vicinity.radius,
                                              paste0("../r_results/predict_correlated_expressed_gene/p1_vicinities_radius_", 
                                                     vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc, 
                                                     "_min-baseMean", min.baseMean, ".bed"))

p2.vicinity.radius = generate_vicinity_radius(p2.dep.regions, 
                                              vicinity.radius,
                                              paste0("../r_results/predict_correlated_expressed_gene/p2_vicinities_radius_", 
                                                     vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc, 
                                                     "_min-baseMean", min.baseMean, ".bed"))

pM.vicinity.radius = generate_vicinity_radius(pM.dep.regions, 
                                              vicinity.radius,
                                              paste0("../r_results/predict_correlated_expressed_gene/pM_vicinities_radius_", 
                                                     vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc, 
                                                     "_min-baseMean", min.baseMean, ".bed"))
```

Calculate Pearson correlation coefficients (PCCs) between NFIA-dependent regions and expressed genes within the assignment areas defined by a fixed radius around the regions:

```{r, include=T}
p1.corr.radius = calc_correlations(p1.vicinity.radius, p1.expr.tss.ranges,
                                   p1.norm.region.counts, p1.norm.gene.counts,
                                   p1.region.sample.names, p1.gene.sample.names,
                                   "p1",
                                   "target",
                                   paste0("../r_results/predict_correlated_expressed_gene/p1_corr_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, ".rds"))

p2.corr.radius = calc_correlations(p2.vicinity.radius, p2.expr.tss.ranges,
                                   p2.norm.region.counts, p2.norm.gene.counts,
                                   p2.region.sample.names, p2.gene.sample.names,
                                   "p2",
                                   "target",
                                   paste0("../r_results/predict_correlated_expressed_gene/p2_corr_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, ".rds"))

pM.corr.radius = calc_correlations(pM.vicinity.radius, pM.expr.tss.ranges,
                                   pM.norm.region.counts, pM.norm.gene.counts,
                                   pM.region.sample.names, pM.gene.sample.names,
                                   "pM",
                                   "target",
                                   paste0("../r_results/predict_correlated_expressed_gene/pM_corr_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, ".rds"))
```

Generate background PCC distributions:

```{r, include=T}
p1.bkgd.radius = calc_correlations(p1.vicinity.radius, p1.expr.tss.ranges,
                                   p1.norm.region.counts, p1.norm.gene.counts,
                                   p1.region.sample.names, p1.gene.sample.names,
                                   "p1",
                                   "background",
                                   paste0("../r_results/predict_correlated_expressed_gene/p1_bkgd_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, ".rds"))
```

Draw the background absolute PCC distribution:

```{r, include=T}
p = ggplot(p1.bkgd.radius) +
  geom_histogram(aes(x = abs_pcc), 
                 binwidth = 0.05) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_deg/plots/p1_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_hist.pdf"),
       plot = p)
```

Calculate and plot the other two background distributions:

```{r, include=T}
p2.bkgd.radius = calc_correlations(p2.vicinity.radius, p2.dep.tss.ranges,
                                   p2.norm.region.counts, p2.norm.gene.counts,
                                   p2.region.sample.names, p2.gene.sample.names,
                                   "p2",
                                   "background",
                                   paste0("../r_results/predict_correlated_deg/p2_bkgd_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, ".rds"))

pM.bkgd.radius = calc_correlations(pM.vicinity.radius, pM.dep.tss.ranges,
                                   pM.norm.region.counts, pM.norm.gene.counts,
                                   pM.region.sample.names, pM.gene.sample.names,
                                   "pM",
                                   "background",
                                   paste0("../r_results/predict_correlated_deg/pM_bkgd_in_radius_",
                                          vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                          "_min-baseMean", min.baseMean, ".rds"))

p = ggplot(p2.bkgd.radius) +
  geom_histogram(aes(x = abs_pcc), 
                 binwidth = 0.05) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_deg/plots/p2_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_hist.pdf"),
       plot = p)

p = ggplot(pM.bkgd.radius) +
  geom_histogram(aes(x = abs_pcc), 
                 binwidth = 0.05) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_deg/plots/pM_bkgd_in_radius_",
                         vicinity.radius, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_hist.pdf"),
       plot = p)
```

Interestingly, the 0.5 peak is not a universal feature of a background distribution of absolute PCCs, and a much bigger proportion of absolute PCCs is concentrated towards the highest end of the scale in p2 and pMN then in p1.

Hereafter, I will use the term "PCC" for "absolute PCC."

Overall, the background distributions suggest, that there will be no significant region-gene associations (because such associations would need to have extremely high PCCs), but let us check this.

Calculate empirical p-values of PCCs:

```{r, include=T}
p1.corr.radius.sign = p1.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, p1.bkgd.radius)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in p1, radius:", nrow(p1.corr.radius.sign %>% filter(sign.padj)), "\n")
```

No significant region-gene associations.

Calculate and correct p-values for PCCs in p2 and pMN:

```{r, include=T}
p2.corr.radius.sign = p2.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, p2.bkgd.radius)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in p2, radius:", nrow(p2.corr.radius.sign %>% filter(sign.padj)), "\n")

pM.corr.radius.sign = pM.corr.radius %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, pM.bkgd.radius)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in pM, radius:", nrow(pM.corr.radius.sign %>% filter(sign.padj)), "\n")
```

Calculate background PCC distributions and empirical p-values in p1, p2 and pMN when using TADs as assignment areas:

```{r, include=T}
p1.bkgd.tad = calc_correlations(p1.vicinity.tad, p1.dep.tss.ranges,
                                p1.norm.region.counts, p1.norm.gene.counts,
                                p1.region.sample.names, p1.gene.sample.names,
                                "p1",
                                "background",
                                paste0("../r_results/predict_correlated_deg/p1_bkgd_in_largest_tad_",
                                       min.tad.size, "-", max.tad.size, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                       "_min-baseMean", min.baseMean, ".rds"))

p2.bkgd.tad = calc_correlations(p2.vicinity.tad, p2.dep.tss.ranges,
                                p2.norm.region.counts, p2.norm.gene.counts,
                                p2.region.sample.names, p2.gene.sample.names,
                                "p2",
                                "background",
                                paste0("../r_results/predict_correlated_deg/p2_bkgd_in_largest_tad_",
                                       min.tad.size, "-", max.tad.size, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                       "_min-baseMean", min.baseMean, ".rds"))

pM.bkgd.tad = calc_correlations(pM.vicinity.tad, pM.dep.tss.ranges,
                                pM.norm.region.counts, pM.norm.gene.counts,
                                pM.region.sample.names, pM.gene.sample.names,
                                "pM",
                                "background",
                                paste0("../r_results/predict_correlated_deg/pM_bkgd_in_largest_tad_",
                                       min.tad.size, "-", max.tad.size, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                                       "_min-baseMean", min.baseMean, ".rds"))

p = ggplot(p1.bkgd.tad) +
  geom_histogram(aes(x = abs_pcc), 
                 binwidth = 0.05) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_deg/p1_bkgd_in_largest_tad_",
                         min.tad.size, "-", max.tad.size, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_hist.pdf"),
       plot = p)

p = ggplot(p2.bkgd.tad) +
  geom_histogram(aes(x = abs_pcc), 
                 binwidth = 0.05) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_deg/p2_bkgd_in_largest_tad_",
                         min.tad.size, "-", max.tad.size, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_hist.pdf"),
       plot = p)

p = ggplot(pM.bkgd.tad) +
  geom_histogram(aes(x = abs_pcc), 
                 binwidth = 0.05) +
  theme_classic()

ggsave(filename = paste0("../r_results/predict_correlated_deg/pM_bkgd_in_largest_tad_",
                         min.tad.size, "-", max.tad.size, "kbp_dep_regions_fdr", fdr, "_min-l2fc", min.l2fc,
                         "_min-baseMean", min.baseMean, "_hist.pdf"),
       plot = p)

p1.corr.tad.sign = p1.corr.tad %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, p1.bkgd.tad)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in p1, TAD:", nrow(p1.corr.tad.sign %>% filter(sign.padj)), "\n")

p2.corr.tad.sign = p2.corr.tad %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, p2.bkgd.tad)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in p2, TAD:", nrow(p2.corr.tad.sign %>% filter(sign.padj)), "\n")

pM.corr.tad.sign = pM.corr.tad %>%
  rowwise() %>%
  mutate(pvalue = calc_empirical_pvalue(abs_pcc, pM.bkgd.tad)) %>%
  ungroup() %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  mutate(fdr = pcc.fdr) %>%
  mutate(sign.padj = (padj < fdr))

cat("Number of significant PCCs ( FDR =", pcc.fdr, ") in pM, TAD:", nrow(pM.corr.tad.sign %>% filter(sign.padj)), "\n")
```

Try to lower the region selection cutoffs and see if any statistically significant region-gene associations (FDR = 5%) emerge. First, re-do the analysis with the strictest cutoffs to have all results in the same format:

```{r, include=T}
# # Set 1 (Joaquina's strictest cutoffs for NFIA-dependent regions)
# fdr = 0.01
# min.l2fc = 2 # Min log2(fold change)
# min.baseMean = 100
# 
# assoc.list.1.radius.p1 = assign_regions_to_genes("p1", "radius")
# 
# assoc.list.1.tad.p1 = assign_regions_to_genes("p1", "tad")
# 
# assoc.list.1.radius.p2 = assign_regions_to_genes("p2", "radius")
# 
# assoc.list.1.tad.p2 = assign_regions_to_genes("p2", "tad")
# 
# assoc.list.1.radius.pM = assign_regions_to_genes("pM", "radius")
# 
# assoc.list.1.tad.pM = assign_regions_to_genes("pM", "tad")
# 
# # Set 2
# fdr = 0.01
# min.l2fc = 2 # Min log2(fold change)
# min.baseMean = 80
# 
# assoc.list.2.radius.p1 = assign_regions_to_genes("p1", "radius")
# 
# assoc.list.2.tad.p1 = assign_regions_to_genes("p1", "tad")
# 
# assoc.list.2.radius.p2 = assign_regions_to_genes("p2", "radius")
# 
# assoc.list.2.tad.p2 = assign_regions_to_genes("p2", "tad")
# 
# assoc.list.2.radius.pM = assign_regions_to_genes("pM", "radius")
# 
# assoc.list.2.tad.pM = assign_regions_to_genes("pM", "tad")
# 
# # Set 3
# fdr = 0.01
# min.l2fc = 2 # Min log2(fold change)
# min.baseMean = 60
# 
# assoc.list.3.radius.p1 = assign_regions_to_genes("p1", "radius")
# 
# assoc.list.3.tad.p1 = assign_regions_to_genes("p1", "tad")
# 
# assoc.list.3.radius.p2 = assign_regions_to_genes("p2", "radius")
# 
# assoc.list.3.tad.p2 = assign_regions_to_genes("p2", "tad")
# 
# assoc.list.3.radius.pM = assign_regions_to_genes("pM", "radius")
# 
# assoc.list.3.tad.pM = assign_regions_to_genes("pM", "tad")
# 
# # Set 4
# fdr = 0.01
# min.l2fc = 2 # Min log2(fold change)
# min.baseMean = 40
# 
# assoc.list.4.radius.p1 = assign_regions_to_genes("p1", "radius")
# 
# assoc.list.4.tad.p1 = assign_regions_to_genes("p1", "tad")
# 
# assoc.list.4.radius.p2 = assign_regions_to_genes("p2", "radius")
# 
# assoc.list.4.tad.p2 = assign_regions_to_genes("p2", "tad")
# 
# assoc.list.4.radius.pM = assign_regions_to_genes("pM", "radius")
# 
# assoc.list.4.tad.pM = assign_regions_to_genes("pM", "tad")
# 
# # Set 5
# fdr = 0.01
# min.l2fc = 2 # Min log2(fold change)
# min.baseMean = 20
# 
# assoc.list.5.radius.p1 = assign_regions_to_genes("p1", "radius")
# 
# assoc.list.5.tad.p1 = assign_regions_to_genes("p1", "tad")
# 
# assoc.list.5.radius.p2 = assign_regions_to_genes("p2", "radius")
# 
# assoc.list.5.tad.p2 = assign_regions_to_genes("p2", "tad")
# 
# assoc.list.5.radius.pM = assign_regions_to_genes("pM", "radius")
# 
# assoc.list.5.tad.pM = assign_regions_to_genes("pM", "tad")

# Set 6 
# fdr = 0.01
# min.l2fc = 1.5 # Min log2(fold change)
# min.baseMean = 20
# 
# assoc.list.6.radius.p1 = assign_regions_to_genes("p1", "radius")
# 
# assoc.list.6.tad.p1 = assign_regions_to_genes("p1", "tad")
# 
# assoc.list.6.radius.p2 = assign_regions_to_genes("p2", "radius")
# 
# assoc.list.6.tad.p2 = assign_regions_to_genes("p2", "tad")
# 
# assoc.list.6.radius.pM = assign_regions_to_genes("pM", "radius")
# 
# assoc.list.6.tad.pM = assign_regions_to_genes("pM", "tad")

# Set 7 
# fdr = 0.01
# min.l2fc = 1 # Min log2(fold change)
# min.baseMean = 20
# 
# assoc.list.7.radius.p1 = assign_regions_to_genes("p1", "radius")
# 
# assoc.list.7.tad.p1 = assign_regions_to_genes("p1", "tad")
# 
# assoc.list.7.radius.p2 = assign_regions_to_genes("p2", "radius")
# 
# assoc.list.7.tad.p2 = assign_regions_to_genes("p2", "tad")
# 
# assoc.list.7.radius.pM = assign_regions_to_genes("pM", "radius")
# 
# assoc.list.7.tad.pM = assign_regions_to_genes("pM", "tad")

# # Set 8 
# fdr = 0.01
# min.l2fc = 0.5 # Min log2(fold change)
# min.baseMean = 20
# 
# assoc.list.8.radius.p1 = assign_regions_to_genes("p1", "radius")
# 
# assoc.list.8.tad.p1 = assign_regions_to_genes("p1", "tad")
# 
# assoc.list.8.radius.p2 = assign_regions_to_genes("p2", "radius")
# 
# assoc.list.8.tad.p2 = assign_regions_to_genes("p2", "tad")
# 
# assoc.list.8.radius.pM = assign_regions_to_genes("pM", "radius")
# 
# assoc.list.8.tad.pM = assign_regions_to_genes("pM", "tad")

# Set 9 
fdr = 0.02
min.l2fc = 0.5 # Min log2(fold change)
min.baseMean = 20

assoc.list.9.radius.p1 = assign_regions_to_genes("p1", "radius")

assoc.list.9.tad.p1 = assign_regions_to_genes("p1", "tad")

assoc.list.9.radius.p2 = assign_regions_to_genes("p2", "radius")

assoc.list.9.tad.p2 = assign_regions_to_genes("p2", "tad")

assoc.list.9.radius.pM = assign_regions_to_genes("pM", "radius")

assoc.list.9.tad.pM = assign_regions_to_genes("pM", "tad")

# Set 10 
fdr = 0.03
min.l2fc = 0.5 # Min log2(fold change)
min.baseMean = 20

assoc.list.10.radius.p1 = assign_regions_to_genes("p1", "radius")

assoc.list.10.tad.p1 = assign_regions_to_genes("p1", "tad")

assoc.list.10.radius.p2 = assign_regions_to_genes("p2", "radius")

assoc.list.10.tad.p2 = assign_regions_to_genes("p2", "tad")

assoc.list.10.radius.pM = assign_regions_to_genes("pM", "radius")

assoc.list.10.tad.pM = assign_regions_to_genes("pM", "tad")

# Set 11 
fdr = 0.04
min.l2fc = 0.5 # Min log2(fold change)
min.baseMean = 20

assoc.list.11.radius.p1 = assign_regions_to_genes("p1", "radius")

assoc.list.11.tad.p1 = assign_regions_to_genes("p1", "tad")

assoc.list.11.radius.p2 = assign_regions_to_genes("p2", "radius")

assoc.list.11.tad.p2 = assign_regions_to_genes("p2", "tad")

assoc.list.11.radius.pM = assign_regions_to_genes("pM", "radius")

assoc.list.11.tad.pM = assign_regions_to_genes("pM", "tad")

# Set 12 
fdr = 0.05
min.l2fc = 0.5 # Min log2(fold change)
min.baseMean = 20

assoc.list.12.radius.p1 = assign_regions_to_genes("p1", "radius")

assoc.list.12.tad.p1 = assign_regions_to_genes("p1", "tad")

assoc.list.12.radius.p2 = assign_regions_to_genes("p2", "radius")

assoc.list.12.tad.p2 = assign_regions_to_genes("p2", "tad")

assoc.list.12.radius.pM = assign_regions_to_genes("pM", "radius")

assoc.list.12.tad.pM = assign_regions_to_genes("pM", "tad")

```
