---
title: "select_diff_regions"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
dyn.load("/home/rstudio/libs/libxml2.so.2")
library(dplyr)
library(BSgenome)
library(BSgenome.Mmusculus.UCSC.mm10)
library(GenomicRanges)
dyn.load("/home/rstudio/libs/libgsl.so.25")
dyn.load("/home/rstudio/libs/libgslcblas.so.0")
library(TFBSTools)
library(tibble)
library(Biostrings)
```

Set FDR:

```{r, include=T}
fdr.regions = 0.01
fdr.matches = 0.05
```

Define a general function for the region selection:

```{r, include=T}
select.regions.by.log2fc = function(deseq.res, count.table, regulation.mode, bed.filename) {
  if (regulation.mode == "up") {
    region.names = rownames(as.data.frame(deseq.res) %>% 
                            dplyr::filter(padj < fdr.regions) %>% 
                            dplyr::filter(log2FoldChange > 0))
  } else if (regulation.mode == "down") {
    region.names = rownames(as.data.frame(deseq.res) %>% 
                            dplyr::filter(padj < fdr.regions) %>% 
                            dplyr::filter(log2FoldChange < 0))
  } else { # any
    region.names = rownames(as.data.frame(deseq.res) %>% 
                            dplyr::filter(padj < fdr.regions))
  }

  bed.content = count.table %>%
    dplyr::select(Geneid,
                  Chr,
                  Start,
                  End,
                  Strand) %>%
    dplyr::mutate(Score = 0) %>%
    dplyr::select(Chr,
                  Start,
                  End,
                  Geneid,
                  Score,
                  Strand) %>%
    dplyr::filter(Geneid %in% region.names)

  write.table(bed.content,
              file = bed.filename,
              sep = "\t",
              quote = F,
              row.names = F,
              col.names = F)
  
  return(GRanges(seqnames = bed.content$Chr,
                 ranges = IRanges(start = bed.content$Start,
                                  end = bed.content$End,
                                  names = bed.content$Geneid),
                 strand = strand(bed.content$Strand)))
}
```

## Select differentially accessible consensus regions

Select consensus regions in p1 that are significantly upregulated, downregulated or significantly regulated in either way:

```{r, include=T}
p1.res = readRDS("../r_results/diff_accessibility/p1_deseq2_results.rds")

p1 = read.delim("../results/select_samples/featureCounts_WT_D7_p1_vs_WT_D11_p1.tsv") %>%
  dplyr::select(-X)

names(p1) = gsub(pattern = ".mLb.clN.bam", replacement = "", x = names(p1))

p1.up.ranges = select.regions.by.log2fc(p1.res, p1, "up", "../r_results/select_diff_regions/p1_upregulated_regions_D11_vs_D7.bed")

p1.down.ranges = select.regions.by.log2fc(p1.res, p1, "down", "../r_results/select_diff_regions/p1_downregulated_regions_D11_vs_D7.bed")

p1.both.ranges = select.regions.by.log2fc(p1.res, p1, "both", "../r_results/select_diff_regions/p1_regulated_regions_D11_vs_D7.bed")
```

Select consensus regions in p2 that are significantly upregulated, downregulated or significantly regulated in either way:

```{r, include=T}
p2.res = readRDS("../r_results/diff_accessibility/p2_deseq2_results.rds")

p2 = read.delim("../results/select_samples/featureCounts_WT_D7_p2_vs_WT_D11_p2.tsv") %>%
  dplyr::select(-X)

names(p2) = gsub(pattern = ".mLb.clN.bam", replacement = "", x = names(p2))

p2.up.ranges = select.regions.by.log2fc(p2.res, p2, "up", "../r_results/select_diff_regions/p2_upregulated_regions_D11_vs_D7.bed")

p2.down.ranges = select.regions.by.log2fc(p2.res, p2, "down", "../r_results/select_diff_regions/p2_downregulated_regions_D11_vs_D7.bed")

p2.both.ranges = select.regions.by.log2fc(p2.res, p2, "both", "../r_results/select_diff_regions/p2_regulated_regions_D11_vs_D7.bed")
```

Select consensus regions in pMN that are significantly upregulated, downregulated or significantly regulated in either way:

```{r, include=T}
pM.res = readRDS("../r_results/diff_accessibility/pMN_deseq2_results.rds")

pM = read.delim("../results/select_samples/featureCounts_WT_D7_pM_vs_WT_D11_pM.tsv") %>%
  dplyr::select(-X)

names(pM) = gsub(pattern = ".mLb.clN.bam", replacement = "", x = names(pM))

pM.up.ranges = select.regions.by.log2fc(pM.res, pM, "up", "../r_results/select_diff_regions/pM_upregulated_regions_D11_vs_D7.bed")

pM.down.ranges = select.regions.by.log2fc(pM.res, pM, "down", "../r_results/select_diff_regions/pM_downregulated_regions_D11_vs_D7.bed")

pM.both.ranges = select.regions.by.log2fc(pM.res, pM, "both", "../r_results/select_diff_regions/pM_regulated_regions_D11_vs_D7.bed")
```

## Scan significantly regulated regions for NFIA motifs

Obtain the sequences of significantly regulated regions in p1, p2 and pMN:

```{r, include=T}
mm10.genome = BSgenome.Mmusculus.UCSC.mm10

p1.da.seq = getSeq(mm10.genome, p1.both.ranges)

p2.da.seq = getSeq(mm10.genome, p2.both.ranges)

pM.da.seq = getSeq(mm10.genome, pM.both.ranges)
```

Prepare background nucleotide frequencies in significantly regulated regions from p1, p2 and pMN:

```{r, include=T}
p1.nc.freq = letterFrequency(p1.da.seq, letters = c("A", "C", "G", "T"))

p1.background = colSums(p1.nc.freq) / sum(p1.nc.freq)

p2.nc.freq = letterFrequency(p2.da.seq, letters = c("A", "C", "G", "T"))

p2.background = colSums(p2.nc.freq) / sum(p2.nc.freq)

pM.nc.freq = letterFrequency(pM.da.seq, letters = c("A", "C", "G", "T"))

pM.background = colSums(pM.nc.freq) / sum(pM.nc.freq)
```

Prepare NFIA motifs to scan:

```{r, include=T}
nfia.hm.pwm.raw = read.delim("../input/NFIA_HUMAN.H11MO.0.C.pwm", 
                             header = F,
                             skip = 1)

names(nfia.hm.pwm.raw) = c("A", "C", "G", "T")

nfia.profile.matrix = t(nfia.hm.pwm.raw)

p1.nfia.hm.pwm = PWMatrix(ID = "NFIA",
                          name = "NFIA",
                          profileMatrix = nfia.profile.matrix,
                          bg = p1.background)

p2.nfia.hm.pwm = PWMatrix(ID = "NFIA",
                          name = "NFIA",
                          profileMatrix = nfia.profile.matrix,
                          bg = p2.background)

pM.nfia.hm.pwm = PWMatrix(ID = "NFIA",
                          name = "NFIA",
                          profileMatrix = nfia.profile.matrix,
                          bg = pM.background)
```

Scan significantly regulated regions (both up- and downregulated) in p1, p2 and pMN for the NFIA binding motif from HOCOMOCO:

```{r, include=T}
# Scan only one strand, as the NFIA motif in HOCOMOCO is palindromic,
# and we do not expect a strand bias in its matches
# (indeed, for p1, I scanned both the "+" and the "-" strand, and the numbers of 
# matched regulated, upregulated and downregulated regions were almost identical)
p1.nfia.hm.sites = searchSeq(p1.nfia.hm.pwm,
                             p1.da.seq,
                             strand = "+",
                             min.score = "80%")

p1.nfia.hm.sites.df = writeGFF3(p1.nfia.hm.sites) %>%
  dplyr::select(-source,
                -feature,
                -frame,
                -attributes) %>%
  mutate(pvalue = unlist(pvalues(p1.nfia.hm.sites, type = "sampling"))) %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  filter(padj < fdr.matches)

p1.matched.regions = unique(p1.nfia.hm.sites.df$seqname)

cat("p1:\n")

cat("Number of matched significantly regulated regions      :", length(unique(p1.nfia.hm.sites.df$seqname)), "\n")

cat("Number of matched significantly upregulated regions    :", 
    sum(p1.matched.regions %in% p1.up.ranges@ranges@NAMES), 
    "(",  round(sum(p1.matched.regions %in% p1.up.ranges@ranges@NAMES) / length(unique(p1.nfia.hm.sites.df$seqname)) * 100, 2), ")%", "\n")

cat("Number of matched significantly downregulated regions  :", 
    sum(p1.matched.regions %in% p1.down.ranges@ranges@NAMES), 
    "(",  round(sum(p1.matched.regions %in% p1.down.ranges@ranges@NAMES) / length(unique(p1.nfia.hm.sites.df$seqname)) * 100, 2), ")%", "\n")

# p2
p2.nfia.hm.sites = searchSeq(p2.nfia.hm.pwm,
                             p2.da.seq,
                             strand = "+",
                             min.score = "80%")

p2.nfia.hm.sites.df = writeGFF3(p2.nfia.hm.sites) %>%
  dplyr::select(-source,
                -feature,
                -frame,
                -attributes) %>%
  mutate(pvalue = unlist(pvalues(p2.nfia.hm.sites, type = "sampling"))) %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  filter(padj < fdr.matches)

p2.matched.regions = unique(p2.nfia.hm.sites.df$seqname)

cat("p2:\n")

cat("Number of matched significantly regulated regions      :", length(unique(p2.nfia.hm.sites.df$seqname)), "\n")

cat("Number of matched significantly upregulated regions    :", 
    sum(p2.matched.regions %in% p2.up.ranges@ranges@NAMES), 
    "(",  round(sum(p2.matched.regions %in% p2.up.ranges@ranges@NAMES) / length(unique(p2.nfia.hm.sites.df$seqname)) * 100, 2), ")%", "\n")

cat("Number of matched significantly downregulated regions  :", 
    sum(p2.matched.regions %in% p2.down.ranges@ranges@NAMES), 
    "(",  round(sum(p2.matched.regions %in% p2.down.ranges@ranges@NAMES) / length(unique(p2.nfia.hm.sites.df$seqname)) * 100, 2), ")%", "\n")

# pMN
pM.nfia.hm.sites = searchSeq(pM.nfia.hm.pwm,
                             pM.da.seq,
                             strand = "+",
                             min.score = "80%")

pM.nfia.hm.sites.df = writeGFF3(pM.nfia.hm.sites) %>%
  dplyr::select(-source,
                -feature,
                -frame,
                -attributes) %>%
  mutate(pvalue = unlist(pvalues(pM.nfia.hm.sites, type = "sampling"))) %>%
  mutate(padj = p.adjust(pvalue, method = "BH")) %>%
  filter(padj < fdr.matches)

pM.matched.regions = unique(pM.nfia.hm.sites.df$seqname)

cat("pMN:\n")

cat("Number of matched significantly regulated regions      :", length(unique(pM.nfia.hm.sites.df$seqname)), "\n")

cat("Number of matched significantly upregulated regions    :", 
    sum(pM.matched.regions %in% pM.up.ranges@ranges@NAMES), 
    "(",  round(sum(pM.matched.regions %in% pM.up.ranges@ranges@NAMES) / length(unique(pM.nfia.hm.sites.df$seqname)) * 100, 2), ")%", "\n")

cat("Number of matched significantly downregulated regions  :", 
    sum(pM.matched.regions %in% pM.down.ranges@ranges@NAMES), 
    "(",  round(sum(pM.matched.regions %in% pM.down.ranges@ranges@NAMES) / length(unique(pM.nfia.hm.sites.df$seqname)) * 100, 2), ")%", "\n")
```

Select significantly upregulated regions with NFIA matches:

```{r, include=T}

```
